server/onerror.ts

```diff
@@ -0,0 +1,138 @@
+import fs from "fs";
+import http from "http";
+import path from "path";
+import Koa, { Context } from "koa";
+import { isNil, escape } from "lodash";
+import env from "@server/env";
+import { InternalError } from "@server/errors";
+
+const isDev = env.ENVIRONMENT === "development";
+const isTest = env.ENVIRONMENT === "test";
+let errorHtmlCache: Buffer | undefined;
+
+const readErrorFile = (): Buffer => {
+  if (!isTest) {
+    return (
+      errorHtmlCache ??
+      (errorHtmlCache = fs.readFileSync(path.join(__dirname, "error.html")))
+    );
+  }
+
+  return (
+    errorHtmlCache ??
+    (errorHtmlCache = fs.readFileSync(
+      path.join(__dirname, "static/index.html")
+    ))
+  );
+};
+
+export default function onerror(app: Koa) {
+  app.context.onerror = function (err: any) {
+    // don't do anything if there is no error.
+    // this allows you to pass `this.onerror`
+    // to node-style callbacks.
+    if (isNil(err)) {
+      return;
+    }
+
+    // wrap non-error object
+    if (!(err instanceof Error)) {
+      let errMsg = err;
+      if (typeof err === "object") {
+        try {
+          errMsg = JSON.stringify(err);
+          // eslint-disable-next-line no-empty
+        } catch (e) {}
+      }
+      const newError = InternalError(`non-error thrown: ${errMsg}`);
+      // err maybe an object, try to copy the name, message and stack to the new error instance
+      if (err) {
+        if (err.name) {
+          newError.name = err.name;
+        }
+        if (err.message) {
+          newError.message = err.message;
+        }
+        if (err.stack) {
+          newError.stack = err.stack;
+        }
+        if (err.status) {
+          newError.status = err.status;
+        }
+        if (err.headers) {
+          newError.headers = err.headers;
+        }
+      }
+      err = newError;
+    }
+
+    const headerSent = this.headerSent || !this.writable;
+    if (headerSent) {
+      err.headerSent = true;
+    }
+
+    // delegate
+    this.app.emit("error", err, this);
```

Comment:
Do we need this for non-500 errors? Looks like the current `errorHandling` middleware for APIs only emits this for 500 errors.

Reply:
It's only need for real internal errors yea, not things like validation. It's actually only used in one spot to call the `requestErrorHandler` that sends errors to sentry so we could call that directly instead of emitting the event